src/devices/machine/gt913.cpp

```diff
@@ -0,0 +1,569 @@
+#include "gt913.h"
+
+DEFINE_DEVICE_TYPE(GT913, gt913_device, "gt913", "Casio GT913F")
+
+gt913_device::gt913_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	h8_device(mconfig, GT913, tag, owner, clock, address_map_constructor(FUNC(gt913_device::map), this)),
+	opcodes_config("opcodes", ENDIANNESS_BIG, 16, 16, 0, address_map_constructor(FUNC(gt913_device::map_opcodes), this)),
+	m_rom(*this, DEVICE_SELF),
+	m_opcodes(*this, ":opcodes"),
+	m_bank(*this, "bank"),
+	m_sound(*this, "gt_sound"),
+	m_kbd(*this, "kbd"),
+	m_uart(*this, "uart"),
+	m_uart_clock(*this, "uart_clock"),
+	m_port(*this, "port%u", 1)
+{
+	m_nmi_vector = 3;
+	m_irq_base = 4;
+}
+
+device_memory_interface::space_config_vector gt913_device::memory_space_config() const
+{	
+	auto spaces = h8_device::memory_space_config();
+	spaces.emplace_back(AS_OPCODES, &opcodes_config);
+	return spaces;
+}
+
+void gt913_device::map(address_map &map)
+{
+//	map.unmap_value_high();
+
+	map(0x0000, 0x7fff).rom();
+	map(0x8000, 0xbfff).bankr("bank");
+	map(0xfac0, 0xffbf).ram(); // CTK-551 zeroes out this range at $0418
+
+	/* ffc0-ffcb: sound */
+	map(0xffc0, 0xffc5).rw(m_sound, FUNC(gt913_sound_hle_device::data_r), FUNC(gt913_sound_hle_device::data_w));
+	map(0xffc6, 0xffc7).w(m_sound, FUNC(gt913_sound_hle_device::command_w));
+	map(0xffca, 0xffcb).r(m_sound, FUNC(gt913_sound_hle_device::status_r));
+
+	/* ffd0-ffd5: key controller */
+	map(0xffd0, 0xffd1).r(m_kbd, FUNC(gt913_kbd_hle_device::read));
+	map(0xffd2, 0xffd3).rw(m_kbd, FUNC(gt913_kbd_hle_device::status_r), FUNC(gt913_kbd_hle_device::status_w));
+
+	/* ffd8-ffdf: timers */
+	map(0xffd8, 0xffd9).rw(FUNC(gt913_device::timer_control_r), FUNC(gt913_device::timer_control_w));
+	map(0xffdc, 0xffdd).w(FUNC(gt913_device::timer_rate0_w));
+	map(0xffdf, 0xffdf).w(FUNC(gt913_device::timer_rate1_w));
+
+	/* ffe0-ffe3: serial */
+	map(0xffe0, 0xffe0).w(FUNC(gt913_device::uart_rate_w));
+	map(0xffe1, 0xffe1).w(m_uart, FUNC(ay31015_device::transmit));
+	map(0xffe2, 0xffe2).rw(FUNC(gt913_device::uart_control_r), FUNC(gt913_device::uart_control_w));
+	map(0xffe3, 0xffe3).r(m_uart, FUNC(ay31015_device::receive));
+
+	/* ffe9-ffea: ADC */
+	map(0xffe9, 0xffe9).rw(FUNC(gt913_device::adc_control_r), FUNC(gt913_device::adc_control_w));
+	map(0xffea, 0xffea).r(FUNC(gt913_device::adc_data_r));
+
+	/* fff0-fff5: I/O ports */
+	map(0xfff0, 0xfff1).rw(FUNC(gt913_device::port_ddr_r), FUNC(gt913_device::port_ddr_w));
+	map(0xfff2, 0xfff2).rw(m_port[0], FUNC(h8_port_device::port_r), FUNC(h8_port_device::dr_w));
+	map(0xfff3, 0xfff3).rw(m_port[1], FUNC(h8_port_device::port_r), FUNC(h8_port_device::dr_w));
+//	map(0xfff4, 0xfff4).nopw(); probably not port 3 DDR - ctk551 writes 0x00 but uses port 3 for output only
+	map(0xfff5, 0xfff5).rw(m_port[2], FUNC(h8_port_device::port_r), FUNC(h8_port_device::dr_w));
+}
+
+void gt913_device::map_opcodes(address_map &map)
+{
+	map(0x0000, 0x7fff).rom().share(":opcodes");
+}
```

Comment:
As soon as you’ve got an absolute tag like this, you know you’re doing it wrong.

Reply:
If you mean the naming of `:opcodes` instead of `opcodes`, that's probably just a syntax misunderstanding on my part.

Reply:
Even if you change it to `opcodes`, it’s making an assumption that the program space always reads from a region of a particular size.  It’s baking in too much of an assumption about the surrounding system.